load-test-framework/src/main/java/com/google/pubsub/clients/common/MetricsHandler.java

```diff
@@ -0,0 +1,222 @@
+        log.error("Unable to initialize MetricsHandler, trying again.", e);
+        executor.execute(this::initialize);
+      } catch (GeneralSecurityException e) {
+        log.error("Unable to initialize MetricsHandler permanently, credentials error.", e);
+      }
+    }
+  }
+
+  private void createMetrics() {
+    try {
+      MetricDescriptor metricDescriptor = new MetricDescriptor()
+          .setType("custom.googleapis.com/cloud-pubsub/loadclient/" + metricName);
+      metricDescriptor.setDisplayName(metricName.toPrettyString())
+          .setDescription(metricName.toPrettyString())
+          .setName(metricDescriptor.getType())
+          .setLabels(Collections.singletonList(new LabelDescriptor()
+              .setKey("client_type")
+              .setDescription("The type of client reporting latency.")
+              .setValueType("STRING")))
+          .setMetricKind("GAUGE")
+          .setValueType("DISTRIBUTION")
+          .setUnit("ms");
+      monitoring.projects().metricDescriptors().create("projects/" + project, metricDescriptor).execute();
+    } catch (Exception e) {
+      log.info("Metrics already exist.");
+    }
+  }
+
+  public synchronized void recordLatency(long latencyMs) {
+    distribution.recordLatency(latencyMs);
+  }
+
+  private void reportMetrics() {
+    CreateTimeSeriesRequest request;
+    synchronized (this) {
+      String now = dateFormatter.format(new Date());
+      request = new CreateTimeSeriesRequest().setTimeSeries(Collections.singletonList(
+          new TimeSeries()
+              .setMetric(new Metric()
+                  .setType("custom.googleapis.com/cloud-pubsub/loadclient/" + metricName)
+                  .setLabels(ImmutableMap.of("client_type", clientType)))
+              .setMetricKind("GAUGE")
+              .setValueType("DISTRIBUTION")
+              .setPoints(Collections.singletonList(new Point()
+                  .setValue(new TypedValue()
+                      .setDistributionValue(new Distribution()
+                          .setBucketCounts(getBucketValues())
+                          .setCount(distribution.getCount())
+                          .setMean(distribution.getMean())
+                          .setSumOfSquaredDeviation(distribution.getSumOfSquareDeviations())
+                          .setBucketOptions(new BucketOptions()
+                              .setExplicitBuckets(new Explicit().setBounds(
+                                  Arrays.asList(ArrayUtils.toObject(LatencyDistribution.LATENCY_BUCKETS)))))))
+                  .setInterval(new TimeInterval()
+                      .setStartTime(now)
+                      .setEndTime(now))))
+              .setResource(monitoredResource)));
+      distribution.reset();
+    }
+    try {
+      monitoring.projects().timeSeries().create("projects/" + project, request).execute();
+    } catch (IOException e) {
+      log.error("Error reporting latency.", e);
+    }
+  }
+
+  List<Long> getBucketValues() {
```

Comment:
No, you need to know the exact buckets, but everything is linked against LatencyDistribution, so it would be impossible to do anything wrong here. The alternative is passing around an entire DistributionValue object, but that would require checking the buckets match up, which seems excessively complicated since we control all the code and only define a single set of buckets.